### PR TITLE
Add retry feature for failed tests

### PR DESCRIPTION
**Pull request by theredfish**
WIP

### CODE REVIEW

**Comment by ilslv** [comment]
@theredfish 

>  I think what I want to implement is very similar to `WhichScenarioFn` in `runner::Basic`

Yes, this functionality should definitely be in `Runner`. And `WhichScenarioFn` is a good starting point. I imaging `RetryableFn` looking something like this:

```rust
enum RetryOrder {
     /// Reties `Scenario` right after the failure.
     Immediately,
     /// Retries `Scenario` in the end. 
     Postponed,
}

pub type RetryableFn = fn(
    &gherkin::Feature,
    Option<&gherkin::Rule>,
    &gherkin::Scenario,
) -> (RetryOrder, usize);
```

Also we should modify the existing `Scenarios` storage to also contain number of retries:

https://github.com/cucumber-rs/cucumber/blob/cf055ac06c7b72f572882ce15d6a60da92ad60a0/src/runner/basic.rs#L1404-L1429

```rust
type RetiesLeft = usize;

type Scenarios = HashMap<
    ScenarioType,
    Vec<(
        Arc<gherkin::Feature>,
        Option<Arc<gherkin::Rule>>,
        Arc<gherkin::Scenario>,
        RetriesLeft,
        RetryOrder,
    )>,
>;
```

So we should re-insert failed `Scenario`s in the front or back based on the `RetryOrder` value. 

> For each retry, launch the scenario again (question : maybe we should add a default timeout between retries with [tokio::time::sleep](https://docs.rs/tokio/latest/tokio/time/fn.sleep.html) for example? But in case of async it could probably be the source of bugs + I'll need to adapt it for `@serial` scenarios).

This should be discussed separately, as currently we are runtime-agnostic and introducing `tokio` dependency would break it. Maybe this would be possible with separate feature enabled, but this is definitely out of the current `PR`s scope.

**Comment by theredfish** [comment]
Thank you for your inputs @ilslv ! I will follow them for the implementation and see how to implement the events 👍

File: src/runner/basic.rs [hunk 1]
```diff
@@ -590,22 +718,87 @@ async fn execute<W, Before, After>(
...
 
-            if fail_fast && scenario_failed {
-                started_scenarios = ControlFlow::Break(());
+            if scenario_failed {
+                if fail_fast {
+                    started_scenarios = ControlFlow::Break(());
+                } else {
+                    // For some reason features.scenarios contains an empty vector for the storage.
+                    // It will fail the rest of the execution. How to access the storage so we can
+                    // read and update the retries?
```
Comment (theredfish) [inline src/runner/basic.rs:742]: @ilslv can you maybe help me to understand this part? I try to read and update the storage. Seems like it's empty at this step (the scenario type is set to `Concurrent` but there is no vector as a value).

**Comment by theredfish** [comment]
Closing in favor of #223

**Pull request closed without merging**